### ISSUE TITLE
Error in Lua Script ! Mods/Scripts/Deathmessage.lua:6: ')' expected near ':'

### ISSUE THREAD

**Issue by TheShade250** [ShadowMario/FNF-PsychEngine#15258]
### What is your question?

i used the code "makeLuaText(tag:String, text:String, width:Int, x:Float, y:Float)" but gives me an error.

Here's a screenshot from my script to creating death message :

How can i do ?

![D__QUENTIN_Jeux_Friday Night Funkin'_PsychEngine 0 7 3 - Copie_mods_scripts_deathmessage lua - Notepad++ 06_08_2024 21_23_10](https://github.com/user-attachments/assets/80869c24-3201-4a54-a596-76ffee5bf2bd)
![Friday Night Funkin'_ Psych Engine 06_08_2024 21_10_36](https://github.com/user-attachments/assets/711c9c7a-a567-4d4f-97fc-da23bc4dd412)
![Friday Night Funkin'_ Psych Engine 06_08_2024 21_22_58](https://github.com/user-attachments/assets/3d8dacc3-375a-43dd-a516-f6dd680503f8)

**Comment by potatex2** [comment]
![image](https://github.com/user-attachments/assets/48728f7b-e75a-45ed-a524-e5580aadbee3)
basically how you use it, also you don't need health to be at -1, the game over inits at `health = 0`
the stuff like `text:String` is just to show the type

**Comment by TheShade250** [comment]
Potatex2, i know but for level mechanics (ex : eye note from Tails.EXE), thet set hp to -1 or more. i try to set a mechanic text if you die, you get the reason of your death. for now, i found how to do this but i'm not so good at coding.

**Comment by TheShade250** [comment]
I reproduced the code. for the moment no message during game over is displayed.

the script :
![D__QUENTIN_Jeux_Friday Night Funkin'_PsychEngine 0 7 3 - Copie_mods_scripts_deathmessage lua - Notepad++ 07_08_2024 00_23_11](https://github.com/user-attachments/assets/bf2a57a0-cf46-44b1-99bd-f99f69c825bb)

**Comment by KillerAvecRage** [comment]
function onGameOver() doesn't allow you doing anything while in game over, use function onGameOverStart() instead

**Comment by TheShade250** [comment]
> function onGameOver() doesn't allow you doing anything while in game over, use function onGameOverStart() instead

Okay, But if I want to put a message indicating the reason for the game over on each way, how do I do it?

**Comment by KillerAvecRage** [comment]
> > function onGameOver() doesn't allow you doing anything while in game over, use function onGameOverStart() instead
> 
> Okay, But if I want to put a message indicating the reason for the game over on each way, how do I do it?

You simply need to put a condition, use a variable which changes depending on the reason why you died

As a demonstration, I'll show you a way to change the death text depending on if the reset button was pressed or health is below 0

```
local pressedReset = false

function onUpdate()
if keyJustPressed('reset') then
pressedReset = true
end
end

function onGameOverStart()
if pressedReset then
makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
else
makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
end
setTextSize('text1', 32)
addLuaText('text1')
end
```

**Comment by TheShade250** [comment]
> > > function onGameOver() doesn't allow you doing anything while in game over, use function onGameOverStart() instead
> > 
> > Okay, But if I want to put a message indicating the reason for the game over on each way, how do I do it?
> 
> You simply need to put a condition, use a variable which changes depending on the reason why you died
> 
> As a demonstration, I'll show you a way to change the death text depending on if the reset button was pressed or health is below 0
> 
> ```
> local pressedReset = false
> 
> function onUpdate()
> if keyJustPressed('reset') then
> pressedReset = true
> end
> end
> 
> function onGameOverStart()
> if pressedReset then
> makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
> else
> makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
> end
> setTextSize('text1', 32)
> addLuaText('text1')
> end
> ```


Okay, i'll try this. 

Could you also add the functions when you die by a mechanic please?

This is for the following notes and/or mechanics.

- Hurt notes (FnF Classic)
- Sawblades (From QT mod)
- Missing Eye note (From VS. Tails.exe)

Once this is done, not only will it help me, but it will help others as well.

**Comment by TheShade250** [comment]
i tried this for adding hurt note death condition but not working for now

![D__QUENTIN_Jeux_Friday Night Funkin'_PsychEngine 0 7 3 - Copie_mods_scripts_deathmessage lua - Notepad++ 11_08_2024 12_16_57](https://github.com/user-attachments/assets/78001996-9413-4500-9891-7bf7af3b8969)

**Comment by KillerAvecRage** [comment]
Pressing a hurt note calls the function noteMiss(id, direction, noteType, isSustainNote)

with those stuff, we are able to identify hurt notes

This giving us
function noteMiss(id, direction, noteType, isSustainNote)
if noteType == 'Hurt Note' then
pressedHurtNote = true
end
end

But this isn't sufficient, a hurt note doesn't necessarily kill so we need to add another condition
A hurt note doesn't deal the same damage as missing a normal note though, so I went in the source code to get the variables and got
0.1 for sustain/long note ; 0.3 for normal note
But there are also gameplay modifiers which influates the variables, such as the instakill on miss modifier or the health loss multiplier modifier so we have to take them into account
these modifiers do not appear in the wiki but don't worry, I already used them before
we're using instakillOnMiss and healthLossMult then
Let's add all of this

local pressedReset = false
local pressedHurtNote = false
local trueHealthLossSus = 0
local trueHealthLossNor = 0

```
function onCreatePost()
trueHealthLossNor = 0.3*healthLossMult
trueHealthLossSus = 0.1*healthLossMult
end

function onUpdate()
if keyJustPressed('reset') then
pressedReset = true
end
end

function onGameOverStart()
if pressedReset then
makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
elseif pressedHurtNote then
makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
else
makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
end
setTextSize('text1', 32)
addLuaText('text1')
end

function noteMiss(id, direction, noteType, isSustainNote)
if noteType == 'Hurt Note' then
if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
pressedHurtNote = true
end
end
end
```

Next thing you want to track is sawblades then, well I can't test it for you, but you can try to find a way to check if the player has been hit by a sawblade and activate a variable when the player does get hit by a sawblade (and kills them)

And if I remember correctly, the tails.exe eye note is an instakill, so it's easier than the hurt note:

```
local pressedReset = false
local pressedHurtNote = false
local trueHealthLossSus = 0
local trueHealthLossNor = 0
local missedEyeNote = false

function onCreatePost()
trueHealthLossNor = 0.3*healthLossMult
trueHealthLossSus = 0.1*healthLossMult
end

function onUpdate()
if keyJustPressed('reset') then
pressedReset = true
end
end

function onGameOverStart()
if pressedReset then
makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
elseif pressedHurtNote then
makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
elseif missedEyeNote then
makeLuaText('text1', 'Missed an eye note', screenWidth, 0, 200)
else
makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
end
setTextSize('text1', 32)
addLuaText('text1')
end

function noteMiss(id, direction, noteType, isSustainNote)
if noteType == 'Hurt Note' then
if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
pressedHurtNote = true
end
end
if noteType == 'Eye Note' then
missedEyeNote = true
end
end
```

There you go, I'll let you find for the sawblade.

**Comment by TheShade250** [comment]
Thanks you KillerAvecRage for helping me ! don't worry, maybe another guy can solve the problem when i create an another issue for sawblades.

**Comment by TheShade250** [comment]
> Pressing a hurt note calls the function noteMiss(id, direction, noteType, isSustainNote)
> 
> with those stuff, we are able to identify hurt notes
> 
> This giving us function noteMiss(id, direction, noteType, isSustainNote) if noteType == 'Hurt Note' then pressedHurtNote = true end end
> 
> But this isn't sufficient, a hurt note doesn't necessarily kill so we need to add another condition A hurt note doesn't deal the same damage as missing a normal note though, so I went in the source code to get the variables and got 0.1 for sustain/long note ; 0.3 for normal note But there are also gameplay modifiers which influates the variables, such as the instakill on miss modifier or the health loss multiplier modifier so we have to take them into account these modifiers do not appear in the wiki but don't worry, I already used them before we're using instakillOnMiss and healthLossMult then Let's add all of this
> 
> local pressedReset = false local pressedHurtNote = false local trueHealthLossSus = 0 local trueHealthLossNor = 0
> 
> ```
> function onCreatePost()
> trueHealthLossNor = 0.3*healthLossMult
> trueHealthLossSus = 0.1*healthLossMult
> end
> 
> function onUpdate()
> if keyJustPressed('reset') then
> pressedReset = true
> end
> end
> 
> function onGameOverStart()
> if pressedReset then
> makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
> elseif pressedHurtNote then
> makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
> else
> makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
> end
> setTextSize('text1', 32)
> addLuaText('text1')
> end
> 
> function noteMiss(id, direction, noteType, isSustainNote)
> if noteType == 'Hurt Note' then
> if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
> pressedHurtNote = true
> end
> end
> end
> ```
> 
> Next thing you want to track is sawblades then, well I can't test it for you, but you can try to find a way to check if the player has been hit by a sawblade and activate a variable when the player does get hit by a sawblade (and kills them)
> 
> And if I remember correctly, the tails.exe eye note is an instakill, so it's easier than the hurt note:
> 
> ```
> local pressedReset = false
> local pressedHurtNote = false
> local trueHealthLossSus = 0
> local trueHealthLossNor = 0
> local missedEyeNote = false
> 
> function onCreatePost()
> trueHealthLossNor = 0.3*healthLossMult
> trueHealthLossSus = 0.1*healthLossMult
> end
> 
> function onUpdate()
> if keyJustPressed('reset') then
> pressedReset = true
> end
> end
> 
> function onGameOverStart()
> if pressedReset then
> makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
> elseif pressedHurtNote then
> makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
> elseif missedEyeNote then
> makeLuaText('text1', 'Missed an eye note', screenWidth, 0, 200)
> else
> makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
> end
> setTextSize('text1', 32)
> addLuaText('text1')
> end
> 
> function noteMiss(id, direction, noteType, isSustainNote)
> if noteType == 'Hurt Note' then
> if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
> pressedHurtNote = true
> end
> end
> if noteType == 'Eye Note' then
> missedEyeNote = true
> end
> end
> ```
> 
> There you go, I'll let you find for the sawblade.

I find for the sawblade missing.

i used the Local dodging = false.

and use the : 
if dodging == false then
makeLuaText('txt1', 'Died due to missing a sawblade. (Press Space to Dodge!)', 'screenWidth, 0, 110)

but they replace all game over message. what can i do ?

**Comment by TheShade250** [comment]
I tried a different thing with another Local :
failedtoDodge = false

and i have this code :
	if characterPlayAnim('boyfriend', 'dodge', false) then
	setProperty('health', -500)
	failedtoDodge = true

idk how to set this. How should I do so that if he doesn't dodge the saw that kills, it displays the message I want to put?

**Comment by KillerAvecRage** [comment]
For that part, you need to take code from the event directly, you can't make local variables interact between scripts, timers however call on every script when it ends, so I could use this to make you a script like this:

```
local pressedReset = false
local pressedHurtNote = false
local trueHealthLossSus = 0
local trueHealthLossNor = 0
local missedEyeNote = false
local missedSawblade = false
local dodging = false

function onCreatePost()
trueHealthLossNor = 0.3*healthLossMult
trueHealthLossSus = 0.1*healthLossMult
end

function onUpdate()
if keyJustPressed('reset') then
pressedReset = true
end
end

function onUpdatePost()
if getPropertyFromClass('flixel.FlxG',"keys.pressed.SPACE") then
if dodging == false then
dodging = true
runTimer("dodgeTimer", 0.2)
end
end
end

function onGameOverStart()
if pressedReset then
makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
elseif pressedHurtNote then
makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
elseif missedEyeNote then
makeLuaText('text1', 'Missed an eye note', screenWidth, 0, 200)
elseif missedSawblade then
makeLuaText('text1', 'Died due to missing a sawblade. (Press Space to Dodge!)', screenWidth, 0, 200)
else
makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
end
setTextSize('text1', 32)
addLuaText('text1')
end

function noteMiss(id, direction, noteType, isSustainNote)
if noteType == 'Hurt Note' then
if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
pressedHurtNote = true
end
end
if noteType == 'Eye Note' then
missedEyeNote = true
end
end

function onTimerCompleted(tag, loops, loopsLeft)
if tag == 'dodgeTimer' then
dodging = false
end

if tag == 'attackTimer' then
if not getProperty("cpuControlled") == true then
if dodging == false then
missedSawblade = true
setProperty('health', -500);
end
end
end
end
```

**Comment by TheShade250** [comment]
> For that part, you need to take code from the event directly, you can't make local variables interact between scripts, timers however call on every script when it ends, so I could use this to make you a script like this:
> 
> ```
> local pressedReset = false
> local pressedHurtNote = false
> local trueHealthLossSus = 0
> local trueHealthLossNor = 0
> local missedEyeNote = false
> local missedSawblade = false
> local dodging = false
> 
> function onCreatePost()
> trueHealthLossNor = 0.3*healthLossMult
> trueHealthLossSus = 0.1*healthLossMult
> end
> 
> function onUpdate()
> if keyJustPressed('reset') then
> pressedReset = true
> end
> end
> 
> function onUpdatePost()
> if getPropertyFromClass('flixel.FlxG',"keys.pressed.SPACE") then
> if dodging == false then
> dodging = true
> runTimer("dodgeTimer", 0.2)
> end
> end
> end
> 
> function onGameOverStart()
> if pressedReset then
> makeLuaText('text1', 'Pressed the reset button', screenWidth, 0, 200)
> elseif pressedHurtNote then
> makeLuaText('text1', 'Pressed a hurt note that killed you', screenWidth, 0, 200)
> elseif missedEyeNote then
> makeLuaText('text1', 'Missed an eye note', screenWidth, 0, 200)
> elseif missedSawblade then
> makeLuaText('text1', 'Died due to missing a sawblade. (Press Space to Dodge!)', screenWidth, 0, 200)
> else
> makeLuaText('text1', 'Health reached 0 or below', screenWidth, 0, 200)
> end
> setTextSize('text1', 32)
> addLuaText('text1')
> end
> 
> function noteMiss(id, direction, noteType, isSustainNote)
> if noteType == 'Hurt Note' then
> if instakillOnMiss or getProperty('health') < trueHealthLossNor or (isSustainNote and getProperty('health') < trueHealthLossSus) then
> pressedHurtNote = true
> end
> end
> if noteType == 'Eye Note' then
> missedEyeNote = true
> end
> end
> 
> function onTimerCompleted(tag, loops, loopsLeft)
> if tag == 'dodgeTimer' then
> dodging = false
> end
> 
> if tag == 'attackTimer' then
> if not getProperty("cpuControlled") == true then
> if dodging == false then
> missedSawblade = true
> setProperty('health', -500);
> end
> end
> end
> end
> ```



Thanks you for helping to solve this. i've had this solution on my eyes but i didn't see this. i'm not so good to create codes like this. I try to learn to creating codes. for now, my mystery has solved. i will put you in credits for helping me and others.

Next time, i will try for myself dor other codes.

Question solved and closed.

**Comment by KillerAvecRage** [comment]
No problem for the help!
